packages/bundle/src/adaptiveCards/Attachment/AdaptiveCardRenderer.js

```diff
@@ -235,18 +235,46 @@ const AdaptiveCardRenderer = ({ actionPerformedClassName, adaptiveCard, disabled
 
   const disabled = disabledFromComposer || disabledFromProps;
 
-  const handleClick = useCallback(
-    ({ target }) => {
+  // TODO: [P2] We should consider using `adaptiveCard.selectAction` instead.
+  // This callback will not listen to "click" or "keypress" event if the component is disabled or does not have "tapAction" prop.
+  const handleClickAndKeyPress = useCallback(
+    event => {
+      const { key, target, type } = event;
+
       // Some items, e.g. tappable text, cannot be disabled thru DOM attributes
-      if (!disabled) {
+      const { current } = contentRef;
+      const adaptiveCardRoot = current.querySelector('.ac-adaptiveCard[tabindex="0"]');
+
+      if (!adaptiveCardRoot) {
+        return console.warn(
+          'botframework-webchat: No Adaptive Card root container can be found, probably on an unsupported Adaptive Card version.'
+        );
+      }
+
+      // For "keypress" event, we only listen to ENTER and SPACEBAR key.
+      if (type === 'keypress') {
+        if (key !== 'Enter' && key !== ' ') {
+          return;
+        }
+
+        event.preventDefault();
+      }
+
+      // We will call performCardAction if either:
+      // 1. We are on the target, or
+      // 2. The event-dispatching element is not interactive
+      if (target !== adaptiveCardRoot) {
         const tabIndex = getTabIndex(target);
 
         // If the user is clicking on something that is already clickable, do not allow them to click the card.
         // E.g. a hero card can be tappable, and image and buttons inside the hero card can also be tappable.
-        if (typeof tabIndex !== 'number' || tabIndex < 0) {
-          tapAction && performCardAction(tapAction);
+        if (typeof tabIndex === 'number' && tabIndex >= 0) {
+          return;
         }
       }
+
+      performCardAction(tapAction);
```

Comment:
Do we need to make sure the tap action isn't undefined?

Reply:
Yes, it is checked in the `onClick` handler below.

I moved the check outside of `useCallback` to reduce the chance it get waste-render. I will put it side-by-side for easier reading.